### PR TITLE
MueLu: Utilities cleanup

### PR DESCRIPTION
**Pull request by cgcgcg**
@trilinos/muelu 

## Motivation
- Refactor `UtilitiesBase`.
- Merge `Utilities_kokkos` into `UtilitiesBase`.

There should be no algorithmic changes except for 
- `GetMatrixDiagonal`,
- `GetMatrixDiagonalInverse`, 
- `GetMatrixOverlappedDiagonal`.

These should now always use the kokkosified variants.

I'll update the PR once #11627 merges.

For review: Only 63c890785739b13f4ddc568cf07e6d7ce13b90d3 and c93c78c21c4a6dc7d0cd45e725b1f116dc8385fe should actually have algorithmic changes. The rest should be purely code refactoring.

### CODE REVIEW

**Comment by GrahamBenHarper** [comment]
related to #11610

**Comment by jhux2** [comment]
Evidence from developer discussion:
```
2023-03-07 discussion:
CG:  Does anyone know if there is any difference between MueLu::Utilities and MueLu::UtilitiesBase? Or could we merge them in principle?
GBH: Is there a difference in how they are templated/instantiated?
     that's what I always assumed was the case
CG:  Not as far as I can tell
CMS: UtilitiesBase = Stuff that's the same regardless of whether we have KokkosRefactor or not
     Utilities = Everything else
GBH: well, then it sounds like it's safe to merge
CG:  Isn't that Utilities vs Utilities_Kokkos?
     Anyway, sounds like it's fine to merge all of that into one file now
CMS: Both Utilities and Utilities_kokkos inherit from UtilitiesBase
CG:  Cause merging will make it a bit easier to see what actually is Kokkosified
CMS: You just need to be careful of algorithmic changes between the Kokkos and non-Kokkos versions
CG:  Yeah, in a first step, I'll just nuke all the ones that just call the base anyway
```

**Comment by cgcgcg** [comment]
Small update with regards to the discussion:
`UtilitiesBase` is everything that does not require partial specialization wrt `EpetraNode`. `Utilities` is derived from `UtilitiesBase` and has all the stuff the needs a special implementation for Epetra.

**Review by jhux2 — APPROVED** [review]
@cgcgcg Looks good to me, thanks!  I have a couple minor questions.  I like how you moved the definitions into the `def` file, that should help with ETI.

File: packages/muelu/src/Operators/MueLu_Maxwell_Utils_def.hpp [hunk 1]
```diff
@@ -89,14 +87,14 @@ namespace MueLu {
     int BCedgesLocal = 0;
     int BCnodesLocal = 0;
     if (useKokkos_) {
-      BCrowsKokkos_ = Utilities_kokkos::DetectDirichletRows(*SM_Matrix_,Teuchos::ScalarTraits<magnitudeType>::eps(),/*count_twos_as_dirichlet=*/true);
+      BCrowsKokkos_ = Utilities::DetectDirichletRows_kokkos(*SM_Matrix_,Teuchos::ScalarTraits<magnitudeType>::eps(),/*count_twos_as_dirichlet=*/true);
```
[the remark this answers is not in the document]
Comment (cgcgcg) [reply packages/muelu/src/Operators/MueLu_Maxwell_Utils_def.hpp:90]: Yeah, the issue is that the only difference between this method and the classical one is the return argument, so we cannot overload. But I will have a look where this method is used, and see if it makes sense to switch all of them to the Kokkos version now.

File: packages/xpetra/src/Utils/ForwardDeclaration/Xpetra_BlockedMap_fwd.hpp [hunk 2]
```diff
@@ -52,7 +52,7 @@ namespace Xpetra {
 }
 
 #ifndef XPETRA_BLOCKEDMAP_SHORT
-#define XPETRA_BLOCKEDAMP_SHORT
+#define XPETRA_BLOCKEDMAP_SHORT
```
[the remark this answers is not in the document]
Comment (cgcgcg) [reply packages/xpetra/src/Utils/ForwardDeclaration/Xpetra_BlockedMap_fwd.hpp:55]: Yeah, that one confused me for a while. It certainly didn't work for `BlockedMap`.

File: packages/muelu/test/unit_tests_kokkos/Utilities_kokkos.cpp [hunk 3]
```diff
@@ -652,32 +654,26 @@ namespace MueLuTests
...
     TEST_EQUALITY(coords, Teuchos::null);
 
     auto map = A->getMap();
     auto mvCoords = Xpetra::MultiVectorFactory<Magnitude, LocalOrdinal, GlobalOrdinal, Node>::Build(map, 1);
     params.set("Coordinates", mvCoords);
-#ifdef HAVE_TPETRA_INST_INT_LONG_LONG
-    coords = Utils_Kokkos::ExtractCoordinatesFromParameterList(params);
+    coords = Utils::ExtractCoordinatesFromParameterList(params);
     TEST_INEQUALITY(coords, Teuchos::null);
-#else
```
[the remark this answers is not in the document]
Comment (cgcgcg) [reply packages/muelu/test/unit_tests_kokkos/Utilities_kokkos.cpp:671]: The non-Kokkos and Kokkos versions had diverged, and it didn't hold true for the non-Kokkos version. But neither of these methods do anything Kokkos-y, so I picked the classical version that seemed more complete.

File: packages/muelu/src/Operators/MueLu_Maxwell_Utils_def.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (cgcgcg) [reply packages/muelu/src/Operators/MueLu_Maxwell_Utils_def.hpp:90]: The ArrayRCP variant `DetectDirichletRows` is used in:
- Vanka smoother
- UnsmooshFactory
- CoalesceDrop (should go away when it's eventually merged with CoalesceDrop_kokkos)
- MaxwellUtils (I'll remove that in my Maxwell refactor)
- unit tests
- Q2Q1 research
I think it's fine to merge as is.

**Pull request merged**